### ISSUE TITLE
Add support for proxy autoconfiguration

### ISSUE THREAD

**Issue by mattico** [rust-lang/rustup#1221]
This would be nice to have. It doesn't look too hard [on Windows](https://msdn.microsoft.com/en-us/library/windows/desktop/aa384095(v=vs.85).aspx) or [on MacOS](https://developer.apple.com/documentation/systemconfiguration/kscpropnetproxiesproxyautoconfigurlstring?language=objc). On linux this seems to be handled by the DE. [Gnome doesn't look too bad](https://stackoverflow.com/questions/7230422/how-to-find-default-proxy-settings-in-ubuntu?rq=1).

**Comment by mattico** [comment]
I'm willing to work on this, once #1183 is done. I'm working on that one, too.

**Comment by mattico** [comment]
A good way to handle this would be to use https://github.com/libproxy/libproxy. Getting it to build cross-platform may be annoying, though.

**Comment by mattico** [comment]
I've started https://github.com/mattico/proxy-config It'll take some time before it's mature enough to use, though.

**Comment by kinnison** [comment]
We use env_proxy these days and I'd assume this is acceptable.  If not, then we should look at enhancing that section of the code.
